### PR TITLE
fix for issue #357

### PR DESCRIPTION
**Pull request by rjaccard**
This solves issue #357

### CODE REVIEW

**Review by Anghul — CHANGES_REQUESTED** [review]
I don't think this fix work when there is multiple servers in the system, as only one server will know the organizer key

**Comment by rjaccard** [comment]
> I don't think this fix work when there is multiple servers in the system, as only one server will know the organizer key

I'm not very familiar with the multiple servers scheme. Only the witnesses and organizers can have a server. So if I add a check for the type of the hub it would be enough, wouldn't be?

File: be1-go/channel/lao/mod.go [hunk 1]
```diff
@@ -283,6 +280,11 @@ func (c *Channel) Publish(publish method.Publish, socket socket.Socket) error {
 
 // processLaoObject processes a LAO object.
 func (c *Channel) processLaoObject(action string, msg message.Message) error {
+	err := verifySenderIsOrganizer(msg, c.organizerPubKey)
+	if err != nil {
+		return err
```
Comment (nkcr) [inline be1-go/channel/lao/mod.go:285]: We always wrap errors ; )

File: be1-go/channel/lao/mod.go [hunk 2]
```diff
@@ -454,22 +456,9 @@ func (c *Channel) processMessageObject(action string, msg message.Message) error
...
-	// Check if the sender of the roll call message is the organizer
-	senderPoint := crypto.Suite.Point()
-	err = senderPoint.UnmarshalBinary(senderBuf)
-	if err != nil {
-		return answer.NewErrorf(-4, keyUnmarshalError, err)
-	}
-
-	if !c.organizerPubKey.Equal(senderPoint) {
-		return answer.NewErrorf(-5, "sender's public key %q does not match the organizer's", msg.Sender)
+		return err
```
Comment (nkcr) [inline be1-go/channel/lao/mod.go:461]: error wrap

File: be1-go/channel/lao/mod.go [hunk 3]
```diff
@@ -530,27 +519,15 @@ func (c *Channel) createChirpingChannel(publicKey string, socket socket.Socket)
...
-
-	if action != expectedAction {
-		return answer.NewErrorf(-4, "invalid action: %s != %s)", action, expectedAction)
-	}
-
-	senderBuf, err := base64.URLEncoding.DecodeString(msg.Sender)
+	err := verifySenderIsOrganizer(msg, c.organizerPubKey)
 	if err != nil {
-		return xerrors.Errorf(keyDecodeError, err)
+		return err
```
Comment (nkcr) [inline be1-go/channel/lao/mod.go:524]: error wrap

File: be1-go/channel/lao/mod.go [hunk 4]
```diff
@@ -711,6 +688,25 @@ func (c *Channel) processRollCallClose(msg messagedata.RollCallClose, socket soc
 	return nil
 }
 
+func verifySenderIsOrganizer(msg message.Message, org kyber.Point) error {
```
Comment (nkcr) [inline be1-go/channel/lao/mod.go:691]: this function could arguably be part of the Channel. Then you already have access to the organizerPukey and must only pass the message.

File: be1-go/channel/lao/mod.go [hunk 5]
```diff
@@ -711,6 +688,25 @@ func (c *Channel) processRollCallClose(msg messagedata.RollCallClose, socket soc
...
+	senderPoint := crypto.Suite.Point()
+	err = senderPoint.UnmarshalBinary(senderBuf)
+	if err != nil {
+		return xerrors.Errorf("failed to unmarshal public key of the sender: %v", msg.Sender)
+	}
+
+	if !senderPoint.Equal(org) {
+		return answer.NewErrorf(-5, "sender's public key %q does not match the organizer's", msg.Sender)
+	}
+	return nil
```
Comment (nkcr) [inline be1-go/channel/lao/mod.go:707]: I would add a newline above

File: be1-go/hub/standard_hub/mod.go [hunk 6]
```diff
@@ -478,13 +478,18 @@ func (h *Hub) createLao(publish method.Publish, laoCreate messagedata.LaoCreate,
...
+	// Check if the sender of create lao message is the organizer
 	senderPubKey := crypto.Suite.Point()
 	err = senderPubKey.UnmarshalBinary(senderBuf)
 	if err != nil {
 		return answer.NewErrorf(-4, "failed to unmarshal public key of the sender: %v", err)
 	}
 
+	if !senderPubKey.Equal(h.pubKeyOrg) {
+		return answer.NewErrorf(-5, "sender's public key %q does not match the organizer's",
+			publish.Params.Message.Sender)
```
Comment (nkcr) [inline be1-go/hub/standard_hub/mod.go:490]:
```suggestion
		return answer.NewErrorf(-5, "sender's public key does not match the organizer's: %q != %q",
			senderPubKey, h.pubKeyOrg)
```

**Comment by Anghul** [comment]
> > I don't think this fix work when there is multiple servers in the system, as only one server will know the organizer key
> 
> I'm not very familiar with the multiple servers scheme. Only the witnesses and organizers can have a server. So if I add a check for the type of the hub it would be enough, wouldn't be?

The biggest problem is that for createLAO, you are using the key known by the hub, but the hub know the key of its owner and not of the organizer. Therefore the witness won't be able to create the copy of the lao needed to keep the necessary state to communicate with the clients connecting to the lao using the witness (I don't know if I'm clear).
Adding a check to let the witness ignore this check isn't the best I think, because it would allow anyone to create a lao on the witnesses.
A possibility would be to make sure that the socket sending the message is an organizer socket, and trust that every message sent by the organizer is correct, but there is still a problem as for now the message are sent to all servers, then treated, so the organizer won't do the check before sending the message further.
I don't have more ideas of how that could be done.

**Pull request closed without merging**